replica/tablets.cc

```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
     co_return std::move(result);
 }
 
+// This sstable set provides access to all the stables in the table, using a snapshot of all
+// its tablets/storage_groups compound_sstable_set:s.
+// The managed sets cannot be modified through tablet_sstable_set, but only jointly read from, so insert() and erase() are disabled.
+class tablet_sstable_set : public sstables::sstable_set_impl {
+    schema_ptr _schema;
+    size_t _log2;
+    // Keep a single (compound) sstable_set per tablet/storage_group
+    std::vector<lw_shared_ptr<sstables::sstable_set>> _sstable_sets;
+    size_t _size;
+    uint64_t _bytes_on_disk;
+
+public:
+    explicit tablet_sstable_set(schema_ptr s, const storage_group_manager& sgm) noexcept
+        : _schema(std::move(s))
+        , _log2(sgm.log2_storage_groups())
+    {
+        _sstable_sets.reserve(sgm.storage_groups().size());
+        for (const auto& sg : sgm.storage_groups()) {
+            if (sg) {
+                auto set = sg->make_sstable_set();
```

Comment:
Wouldn't sg maintain an sstable set? do we have to create a new one?

Each time a tablet migrates we have to make_sstable_set() on all tablets, where we can just pick up an existing set.

Reply:
The name sstable_set is a bit misleading since the storage group returns a compound sstable set that just collects the underlying sstable sets using a shallow copy so it's relatively cheap.
That said, I think we can keep a pre-calculated lw_shared_ptr<sstable_set> per storage_group and just grab a reference to it, but we'll need to rebuild it when any of its underlying compaction_group's sstable_set:s is updated - this makes a tree-like refresh process during cache update where the leaf-level sstable sets are replaced and the compound sstable sets on top of them need to updated, up to the top level which is this tablet_sstable_set (in case tablets are used, otherwise it would be the single compaction_group compound_sstable_set which shortens this tree).

In any case, I suggest we save this as a followup since to keep this change relatively conservative and test the hell out of it while working on the next stage.